### PR TITLE
builtin: str.Number?() now matches numbers with underscores

### PR DESCRIPTION
**Pull request by frroossst**
image

```
_123         => does NOT match
1____2       => does match
1___.___3    => does match
```

### CODE REVIEW

File: builtin/string.go [hunk 1]
```diff
@@ -349,7 +349,7 @@ func string_NumberQ(this Value) Value {
 }
 
 // benchmark shows Suneido regex is faster than Go regexp for this
-var numberPat = regex.Compile(`\A[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d\d?)?\Z`)
+var numberPat = regex.Compile(`\A[+-]?(\d+([\.][\d_]*)?|[\.][\d_]+)([eE][+-]?\d\d?)?\Z`)
```
Comment (apmckinlay) [inline builtin/string.go:352]: do we need the square brackets around `\.` ?

**Comment by apmckinlay** [comment]
We should probably add a few test cases to TestNumberPat in x_test.go

File: builtin/string.go (hunk 1 above)
Comment (frroossst) [reply builtin/string.go:352]: checking in with regexper doesn't look like they are needed, have removed them

**Comment by frroossst** [comment]
Okay so write a couple of tests, and definitely broke a few things, will push a fix soon.

**Comment by frroossst** [comment]
Wrote a very complicated looking regex, maybe could be further simplified.

`\A[+-]?(\d+([_\d]+)*([\.]([_\d]+)*)?|([\.]\d+([_\d]+)*))([eE][+-]?\d+)?\Z`

image

**Comment by apmckinlay** [comment]
`[\.]` is redundant, either `\1` or `[.]` is sufficient.

The old regex was "one pass", whereas the new one isn't. That's not a huge issue, but it will make it slower. You can see this with Suneido.Regex(...).Disasm() and look for OnePass at the beginning.

You made some changes that seem unrelated to underscores. The exponent now takes only one or two digits, which doesn't match lexer number.

It appears to allow `1._2` but not `._2`
Lexer number currently allows `._2` although we could change that.

Maybe we should just use Lexer instead of a regular expression and avoid two separate implementations with potentially slightly different behavior.

**Comment by apmckinlay** [comment]
It just requires some setup to use Lexer.
It might not be too bad, we should try it.

jSuneido doesn't use a regex, it basically duplicates the lexer code,
which is also ugly.

**Comment by apmckinlay** [comment]
This seems to work and benchmarks almost twice as fast:
```
func isNumber(s string) bool {
	lexer := lexer.NewLexer(s)
	item1 := lexer.Next()
	if item1.Token == tokens.Add || item1.Token == tokens.Sub {
		item1 = lexer.Next()
	}
	item2 := lexer.Next()
	return item1.Token == tokens.Number && item2.Token == tokens.Eof
}
```

But I needed to change lexer number to not accept just "0x"
```
lxr.matchWhile(IsHexDigit)
if lxr.si == start + 2 {
	lxr.si = start + 1 // if only 0x just absorb the 0
}
```

**Comment by frroossst** [comment]
haha, I was just doing that, you beat me to it lol. 

Out of curiosity do you benchmark Go or gSuneido? I don't imagine there being much of a difference, in terms of relative performance.

**Comment by apmckinlay** [comment]
I use the Go benchmarking for small micro benchmarks.
You can find various benchmarks in the _test.go files.
In gSuneido things like the call overhead and interpreter would swamp what
you were measuring.

I'll benchmark in gSuneido for larger scale stuff
e.g. with Timer(secs: 5) { ... }

**Comment by apmckinlay** [comment]
Sorry for beating you to it. I wasn’t sure how hard it would be.

**Comment by frroossst** [comment]
haha no worries on that, all good.

**Comment by apmckinlay** [comment]
I realize one difference is that the lexer version will accept hex (0x...)
whereas the old version didn't.
It's probably ok that it does, I'll just have to make jSuneido the same.

**Comment by apmckinlay** [comment]
Another question is whether hex numbers should allow underscores?
I would probably say yes. (e.g. Go allows it)
At least if we use the lexer we only have to handle it in one place.

**Comment by frroossst** [comment]
I'd say any and all numbers allow underscores. Like, `1_000e1_2` should also be valid. Basically.
```
for i in number:
    if i is '_':
        pass
    else:
        parseNum()
```

File: builtin/string.go [hunk 2]
```diff
@@ -350,7 +352,7 @@ func string_NumberQ(this Value) Value {
 }
 
 // benchmark shows Suneido regex is faster than Go regexp for this
-var numberPat = regex.Compile(`\A[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d\d?)?\Z`)
```
Comment (apmckinlay) [inline builtin/string.go:353]: I thought we were going to switch to the lexer approach?

Comment (frroossst) [reply builtin/string.go:353]: Sorry, discrepancies between my home computer and work computer.

**Comment by frroossst** [comment]
I was wondering what the `string_NumbericQ` method was for, and if that also needs to be changed. From the looks of it I don't think that needs to be changed, but could be wrong.

**Comment by apmckinlay** [comment]
I think it's ok. It's not trying to recognize numbers, just character classes, it goes with Alpha? and AlphaNum? (AlphaNum being short for AlphaNumeric)

File: builtin/string.go [hunk 3]
```diff
@@ -346,11 +348,19 @@ var _ = method(string_NumberQ, "()")
...
 	// see also Lexer.number
-	return SuBool(numberPat.Matches(ToStr(this)))
+	lexer := lexer.NewLexer(ToStr(this))
+	item := lexer.Next()
+	if item.Token == tok.Add || item.Token == tok.Sub {
+		item = lexer.Next()
+	}
+	if item.Token == tok.Number || item.Token == tok.Eof {
+		return True
+	}
```
Comment (apmckinlay) [inline builtin/string.go:358]: wouldn't this accept Add, Eof ?

Comment (frroossst) [reply builtin/string.go:358]: yes, which is not what we want. fixing it now

Comment (frroossst) [reply builtin/string.go]: I think I fixed it

1. Consume +/- if they exist
2. Consume number (takes care of _, e, .)
3. Match `Eof` after the number to make sure we have gotten the number
4. If not return False

File: builtin/string.go [hunk 4]
```diff
@@ -346,11 +348,22 @@ var _ = method(string_NumberQ, "()")
...
+		if item.Token == tok.Eof {
+			return True
+		}
+	}
+	return False
 }
 
 // benchmark shows Suneido regex is faster than Go regexp for this
-var numberPat = regex.Compile(`\A[+-]?(\d+(\.\d*)?|\.\d+)([eE][+-]?\d\d?)?\Z`)
+var numberPat = regex.Compile(`\A[+-]?(\d+([_\d]+)*([\.]([_\d]+)*)?|([\.]\d+([_\d]+)*))([eE][+-]?\d+)?\Z`)
```
Comment (apmckinlay) [inline builtin/string.go:366]: We can delete the comment and numberPat.

Comment (frroossst) [reply builtin/string.go:366]: yes, I still haven't changed the tests. I'll do it with that push

**Comment by apmckinlay** [comment]
Is this ready to go?

**Comment by frroossst** [comment]
I believe so, yes. If you'd want to take one final look

File: builtin/string.go [hunk 5]
```diff
@@ -346,12 +348,20 @@ var _ = method(string_NumberQ, "()")
...
+	if item.Token == tok.Add || item.Token == tok.Sub {
+		item = lexer.Next()
+	}
+	if item.Token == tok.Number {
+		item = lexer.Next()
+		if item.Token == tok.Eof {
+			return True
+		}
+	}
+	return False
```
Comment (apmckinlay) [inline builtin/string.go:362]: It's minor, but I would avoid the extra nesting and keep this more "straight line".

```
	if item.Token != tok.Number {
		return False
	}
	item = lexer.Next()
	return item.Token == tok.Eof
```

File: builtin/x_test.go [hunk 6]
```diff
@@ -10,19 +10,17 @@ import (
...
-			assert.True(numberPat.Matches("+" + s))
-			assert.True(numberPat.Matches("-" + s))
-			assert.False(numberPat.Matches("x" + s))
-			assert.False(numberPat.Matches(s + "x"))
-		}
+		// write test cases to test string_NumberQ function
+		// don't test regex
+		// func string_NumberQ(this runtime.Value) runtime.Value
+		// convert string to Value type
+		assert.T(t).This(string_NumberQ(SuStr(s))).Is(SuBool(expected))
```
Comment (apmckinlay) [inline builtin/x_test.go:23]: remove comments ?

File: builtin/x_test.go [hunk 7]
```diff
@@ -10,19 +10,17 @@ import (
 
 	"github.com/apmckinlay/gsuneido/util/assert"
 	"github.com/apmckinlay/gsuneido/util/regex"
+	. "github.com/apmckinlay/gsuneido/runtime"
 )
 
 func TestNumberPat(t *testing.T) {
```
Comment (apmckinlay) [inline builtin/x_test.go:16]: rename to TestNumberQ

File: builtin/x_test.go [hunk 8]
```diff
@@ -47,6 +45,30 @@ func TestNumberPat(t *testing.T) {
...
+	test("1._0_", true)
+	test("1_.0_", true)
+	test("1.0__", true)
+	test("1.__0", true)
+	test("1.__0_", true)
+	test("1.__0__", true)
+	test("1.0__0", true)
+	test("__1.0__0_", false)
+	test("__1.0__0__", false)
+
```
Comment (apmckinlay) [inline builtin/x_test.go:71]: test cases for hex ?

File: builtin/string.go (hunk 5 above)
Comment (frroossst) [reply builtin/string.go:362]: The function returns a value and the `==` would evaluate to a bool, should I just wrap the return value in a SuBool?

**Comment by apmckinlay** [comment]
Yes, sorry, missed that.

**Pull request merged**